runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/DeferredHeaders.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.http
+
+import aws.smithy.kotlin.runtime.http.util.*
+import kotlinx.coroutines.Deferred
+
+/**
+ * Immutable mapping of case insensitive HTTP header names to list of [String] values.
+ */
+public interface DeferredHeaders : DeferredValuesMap<String> {
+    public companion object {
+        public operator fun invoke(block: DeferredHeadersBuilder.() -> Unit): DeferredHeaders = DeferredHeadersBuilder()
+            .apply(block).build()
+
+        /**
+         * Empty [Headers] instance
+         */
+        public val Empty: DeferredHeaders = EmptyDeferredHeaders
+
+        public suspend fun DeferredHeaders.toHeaders(): Headers = when (this) {
```

Comment:
**Quesiton**: What's the use case for this conversion? It seems...odd.

Reply:
It's mainly a convenience function for calling `.await()` on all of the `Deferred` values and converting them to `String`

Reply:
It's ok I suppose but it was harder to follow the implementation in the chunked implementation. This should have some documentation about what it's doing and probably be given an `InternalApi` annotation.